### ISSUE TITLE
Filtering by endpoint paths or tags?

### ISSUE THREAD

**Issue by er1c** [ghostbuster91/sttp-openapi-generator#46]
Original source: https://twitter.com/ericpeters/status/1394315772793483271

Example yaml: https://github.com/github/rest-api-description/blob/main/descriptions/api.github.com/api.github.com.yaml

# Use Case

Two possibilities:

1. Like this example, GitHub has a broken YAML file, where some of their API endpoints are complete, others are not.
2. You don't need (or want) a complete client generated for some of these large enterprise REST services.  I don't care about the billing, scim, code-scanning, etc - in this example the API has some tags, or maybe the endpoint prefixes?

```
tags:
- name: actions
  description: Endpoints to manage GitHub Actions using the REST API.
- name: activity
  description: Activity APIs provide access to notifications, subscriptions, and timelines.
- name: apps
  description: Information for integrations and installations.
- name: billing
  description: Monitor charges and usage from Actions and Packages.
- name: checks
  description: Rich interactions with checks run by your integrations.
- name: classroom
  description: Endpoints that give information about Classroom-related data.
- name: code-scanning
  description: Retrieve code scanning alerts from a repository.
- name: codes-of-conduct
  description: Insight into codes of conduct for your communities.
- name: emojis
  description: List emojis available to use on GitHub.
- name: gists
  description: View, modify your gists.
- name: git
  description: Raw Git functionality.
- name: gitignore
  description: View gitignore templates
- name: interactions
  description: Owner or admin management of users interactions.
- name: issues
  description: Interact with GitHub Issues.
- name: licenses
  description: View various OSS licenses.
- name: markdown
  description: Render Github flavored markdown
- name: meta
  description: Endpoints that give information about the API.
- name: migrations
  description: Move projects to or from GitHub.
- name: oauth-authorizations
  description: Manage access of OAuth applications
- name: orgs
  description: Interact with GitHub Orgs.
- name: projects
  description: Interact with GitHub Projects.
- name: pulls
  description: Interact with GitHub Pull Requests.
- name: rate-limit
  description: Check your current rate limit status
- name: reactions
  description: Interact with reactions to various GitHub entities.
- name: repos
  description: Interact with GitHub Repos.
- name: scim
  description: Provisioning of GitHub organization membership for SCIM-enabled providers.
- name: search
  description: Look for stuff on GitHub.
- name: secret-scanning
  description: Retrieve secret scanning alerts from a repository.
- name: teams
  description: Interact with GitHub Teams.
- name: users
  description: Interact with and view information about users and also current user.
```

**Comment by ghostbuster91** [comment]
Hi,

Thanks for trying out the generator :) 

This is an interesting case and I believe that in general users won't be interested in generating code for the whole GitHub specification. Even if everything worked, the compilation times would be bloated up unnecessarily. This means that there should be a way to filter only a subset of operations. I think that relying on tags is both simple and enough. 

Having said that a few things need to be done to implement such an approach. 
- [ ] Users can have multiple open-api specifications, but I doubt that all of them should be filtered using the same set of tags. A fine-grained mechanism of passing input parameters would be needed. I already started working on it in https://github.com/ghostbuster91/sttp-openapi-generator/pull/48
- [ ] Even if we filter out only some subset of operations a generated set of classes will contain request/response objects from the whole spec. A mechanism to only generate model classes that are referenced by operations is necessary.

**Comment by ghostbuster91** [comment]
I've just pushed v0.2.0 with the mentioned changes (fine-grained loading and rendering only referenced schemas). Still, there is no mechanism to tell the generator to filter out some subset of operations. But this can be done totally outside of the generator, before even running it.

You could use open-api parser to perform such filtering in a simple task in your build.sbt
As the open-api parser is very javaish and full of nulls I would recommend using the tiny wrapper I made for the generator. It isn't perfect for that use-case but still better than the original one. 

```
import _root_.io.github.ghostbuster91.sttp.client3.SbtLogAdapter
import _root_.io.github.ghostbuster91.sttp.client3.SttpOpenApiCodegenPlugin._
import _root_.io.github.ghostbuster91.sttp.client3.openapi.SafeOpenApiParser
import _root_.io.swagger.v3.oas.models.PathItem.HttpMethod
import _root_.sttp.model.Method

import _root_.io.swagger.v3.core.util.Yaml

val sampleIntTask = taskKey[Unit]("A sample int task.")

lazy val petstoreApi: Project = project
  .in(file("petstore-api"))
  .enablePlugins(SttpOpenApiCodegenPlugin)
  .settings(
    sampleIntTask := {
      val logAdapter= new SbtLogAdapter(streams.value.log)
      val parser = new SafeOpenApiParser(logAdapter, List.empty)
      val result = parser.parse(IO.read(file("/workspace/sttp-openapi-example/petstore-api/src/main/resources/com/example/sttp/pet_store.yaml")))
      result.map { spec =>
        spec.paths.values.foreach { pathItem =>
          val operationsToDelete = pathItem.operations.filter { case (_, operation) => !operation.tags.toList.flatten.contains("actions") }
          operationsToDelete.foreach { case(method, _) =>
            pathItem.unsafe.operation(sttpMethodToOpenApiMethod(method), null)
          }
        }
        IO.write(file("/workspace/sttp-openapi-example/petstore-api/src/main/resources/com/example/sttp/pet_store_2.yaml"),Yaml.pretty(spec.unsafe))
      }
    },
    sttpOpenApiInput := Seq(Input.SingleFile(file("/workspace/sttp-openapi-example/petstore-api/src/main/resources/com/example/sttp/pet_store_2.yaml"),pkg ="com.example"))
  )

def sttpMethodToOpenApiMethod(method: Method) : HttpMethod = {
  method match {
    case Method.POST => HttpMethod.POST
    case Method.GET => HttpMethod.GET
    case Method.PUT => HttpMethod.PUT
    case Method.PATCH => HttpMethod.PATCH
    case Method.DELETE => HttpMethod.DELETE
    case Method.HEAD => HttpMethod.HEAD
    case Method.OPTIONS => HttpMethod.OPTIONS
    case Method.TRACE => HttpMethod.TRACE
  }
}
```

I think that such an approach is better as it respects the separation of concerns (in other words, the generator knows only about generating, and filtering is done somewhere else).

Although the filtering seems to work, I was unable to generate client code for a subset of operations that have `action` tag. If you would like to help, minimizing that subset to the smallest possible example would be extremely helpful. 

To sum up, I think that it should be the way to go, however we would need a better parser to make it a pleasant experience for the users. It is something which I have on my endless list of todos ;) 

Let me know what do you think about it.

**Comment by ghostbuster91** [comment]
closing due to inactivity
